README.md
# Finnish Meteorological Institute Probabilistic Precipitation Nowcasting system (FMI-PPN)
FMI-PPN is a modular weather-radar-based nowcasting system built for research and operational usage.

## Subsystems
### Precipitation motion
Current implementation of FMI-PPN uses Lucas-Kanade optical flow method to estimate precipitation movement from radar measurements.

### Nowcasting
Currently FMI-PPN uses [pysteps](https://pysteps.github.io) to generate ensemble nowcasts.

## Known issues and limitations
- Parameter `SEED` must be `None` or an integer between `0` and `2**32 - 1`. This is a limitation in `numpy.random`.
- OpenMPI conflicts with dask when both are installed, leading to significant decrease FMI-PPN performance. Workaround is to set OpenMPI use only one thread. In Linux you can set environment variable `OMP_NUM_THREADS=1`.

## Usage
### Installation
1. Install [conda](https://conda.io/en/latest/) (Miniconda is recommended)
2. Clone this repository
3. Change directory to FMI-PPN folder
4. Setup conda environment: `$ conda env create -f environment.yml`
5. Modify pystepsrc file: Add your data source configuration under `data_sources` (see pysteps documentation for details)
6. Replace the value for `DOMAIN` parameter under `defaults` with your data source configuration's name
7. Activate conda environment: `$ conda activate fmippn`
8. Run FMI-PPN with default settings: `$ python run_ppn.py`

### Running FMI-PPN
Before running FMI-PPN, you should configure pysteps (via `pystepsrc` file) and PPN by adding your parametrisations to `ppn_config.py`.

How to run:
1. Activate your conda environment
2. Run FMI-PPN with your settings: `$ python run_ppn.py -c your_config_parametrisation`

## Configuration
### Adding new parametrisations

### Parametrisations
Parameter|Explanation|Default value
----|----|----
`DBZ_MIN`|Minimum value used in pystep cascade calculations, when calculation is performed for dBZ units|`-10`
`DBZ_THRESHOLD`|Thresholding value used by pysteps, when calculation is performed for dBZ units|`-10`
`DOMAIN`|Data source used from pystepsrc|`fmi`
`ENSEMBLE_SIZE`|Number of ensemble members|`25`
`FFT_METHOD`|FFT method used in pysteps calculations|`pyfftw`
`FIELD_VALUES`|Select the units to store the nowcast (before scaling). Valid units are `dbz` for dBZ and `rrate` for mm/h.|`dbz`
`GENERATE_DETERMINISTIC`|Calculate extrapolation-only nowcast|`True`
`GENERATE_ENSEMBLE`|If `True`, then calculate ensemble members|`True`
`GENERATE_UNPERTURBED`|Calculate nowcast using pysteps, but without noise|`False`
`KMPERPIXEL`|Pixel size in kilometers|`1.0`
`LOG_FOLDER`|Path where log files should be stored|`../logs`
`LOG_LEVEL`|Logging level used by [Python Logging Module](https://docs.python.org/3/library/logging.html#levels)|`20`
`MAX_LEADTIME`|How long your nowcast will be (in minutes)|`120`
`NOWCAST_TIMESTEP`|Timestep between consecutive nowcast images|`5`
`NUM_CASCADES`|How many cascade levels are used in cascade decomposition by pysteps|`6`
`NUM_PREV_OBSERVATIONS`|Number of previous observations used in optical flow calculation|`3`
`NUM_TIMESTEPS`|How many timesteps will be calculated in nowcasts (If this setting is `None`, this value is automatically calculated based on `MAX_LEADTIME` and `NOWCAST_TIMESTEP` parameters) |`None`
`NUM_WORKERS`|Number of worker threads used in parallel computing|`6`
`OPTFLOW_METHOD`|Optical flow method (see pysteps docs)|`lucaskanade`
`OUTPUT_PATH`|If not `None`, then use this path to store output instead of setting in pysteprc|`None`
`OUTPUT_TIME_FORMAT`|Python datetime format for showing timestamps|`%Y-%m-%d %H:%M:%S`
`REGENERATE_PERTURBED_MOTION`|Re-calculate motion for each ensemble member (requires `SEED != None`) |`False`
`R_MIN`|Minimum value used in pystep cascade calculations, when calculation is performed for mm/h units|`0.1`
`R_THRESHOLD`|Thresholding value used by pysteps, when calculation is performed for mm/h units|`0.1`
`SCALER`|Scaling coefficient for output|`100`
`SCALE_ZERO`|Value for "0" after scaling the output. Setting to `"auto"` or `None` uses the minimum value found in data before scaling.|`auto`
`SEED`|Seed parameter for random number generation. Use `None` for unseeded nowcasts.|`None`
`STORE_DETERMINISTIC`|Write "deterministic" nowcast in output file|`True`
`STORE_ENSEMBLE`|Write all ensemble members in output file|`True`
`STORE_MOTION`|Write optical flow motion field in output file|`True`
`STORE_PERTURBED_MOTION`|Write optical flow motion for each ensemble member in output file|`True`
`STORE_UNPERTURBED`|Write "unperturbed" nowcast in output file|`True`
`VALUE_DOMAIN`|Choose if nowcasting is performed for data in dBZ or mm/h units. Valid parameters are `dbz` for dBZ and `rrate` for mm/h.|`dbz`
`VEL_PERT_KWARGS`|Parameters for velocity perturbation (see pysteps docs). Set this parameter to `None` to use default pysteps values.|`{'p_par': [2.20837526, 0.33887032, -2.48995355], 'p_perp': [2.21722634, 0.32359621, -2.57402761]}`
`VEL_PERT_METHOD`|Velocity perturbation method used in pysteps|`bps`
`WRITE_LOG`|If `True`, then generate a log file|`False`
`ZR_A`|Value for coefficient _a_ in R(Z) relation|`223.0`
`ZR_B`|Value for coefficient _b_ in R(Z) relation|`1.53`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

This project DOES NOT adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). Instead, version numbering will be a running number. Each major number increases when a dependency module is upgraded, added, or removed (e.g. pysteps). Each minor number increases when a research version is updated, either by RAS or PAK.

## [v2.1] - 2019-12-20
### Added
- New parametrisations and configurations for operational use
- Usage and Configuration descriptions to readme

### Changed
- Now `pystepsrc` file is tracked, previously it was ignored
- Renamed "dev" parametrisation to "test"
- Updated default parameters: now generates 25 ensemble members for up to 120 minutes
- Updated docstrings
- Updated .gitignore
- Improved markdown syntax usage in changelog and readme

### Removed
- Parametrisations used in research

### Fixed
- Dangerous default argument value in `utils.store_timeseries()`
- Removed unused import from `utils`
- Version tag links in changelog

## [v2.0] - 2019-09-02
### Added
- Changelog
- New dependency: pysteps (1.0.0)
- Now it is possible to set motion perturbation parameters
- Default motion perturbation parameters for Lucas-Kanade method
- STEPS nowcasts can now be calculated in dBZ, no need to convert to rain rate first
- Functionality for regenerating perturbed motion fields (for a given `SEED`)
- Projection metadata is now stored in output file
- Configuration parameters and their values are now stored in output file
- Added new configuration parameters for controlling program flow
- Added configuration for heavy rainfall event in Helsinki on 23 Aug 2019

### Changed
- Renamed readme.md to README.md
- Deterministic nowcast uses now extrapolation
- Old deterministic nowcast is renamed as "unperturbed nowcast"
- `utcnow_floored()` function was moved to `utils` module
- Edited "dev" and "verification" parameters a bit

### Removed
- Extra dependencies that were unused or no longer needed

## [v1.1] - 2019-05-06
### Added
- New parameters: `SEED`, `ZR_A`, `ZR_B`, `SCALE_ZERO`, `SCALER`, `FIELD_VALUES`
- Now it is possible to set random number generator seed via `SEED` parameter
- Now R(Z)-relation coefficients _a_ and _b_ can be set via parameters `ZR_A` and `ZR_B`
- Nowcasts can be stored either in dBZ or mm/h units. This is chosen via `FIELD_VALUES` parameter
- Scaling parameter in output file can now be set via parameter `SCALER`
- Scaling offset in output file can now be set via `SCALE_ZERO` parameter
- Program uses nowcast data array's minimum non-NaN value as scaling offset, unless given via configuration parameters
- Parameters `SEED` and `FIELD_VALUES` are stored in output file`s metadata

### Changed
- "Valid for" attribute in output file is now an integer (was: string)
- `OUTPUT_TIME_FORMAT` parameter no longer affects the "Valid for" attribute

## [v1.0] - 2019-01-22
- First prototype version given to operational testing

[v2.1]: https://github.com/fmidev/fmippn/compare/v2.0...v2.1
[v2.0]: https://github.com/fmidev/fmippn/compare/v1.1...v2.0
[v1.1]: https://github.com/fmidev/fmippn/compare/v1.0...v1.1
[v1.0]: https://github.com/fmidev/fmippn/releases/tag/v1.0



ppn_config.py
"""Parameter configuration for FMI-PPN

Dictionary `defaults` contains default parameter settings for FMI-PPN.

Other dictionaries should contain non-default parameters. Program should
update defaults with non-default dictionary.

Utility method `get_params(name)` returns a dictionary.

Adding new parametrizations:
  1. Create a new dictionary object
  2. Add new parameter values in the new dictionary
  3. Update the lookup dictionary inside get_params() method
"""
import logging


def get_params(name):
    """Utility function for easier access to wanted parametrizations.

    Two parametrisations should be provided: `defaults` and `test`.
    `defaults` contains all possible parameters and default values for them.
    The default values are used unless explicitly overriden.

    `test` contains parametrisations for short nowcast that generates all
    different nowcasts. Suitable for quickly testing different parameters and
    for development purposes.

    Input:
        name -- name of the parametrization dictionary

    Return dictionary of parameters for overriding defaults. Non-existing
    names will return an empty dictionary.
    """
    # Lookup dictionary
    names = {
        'defaults': defaults,
        'test': test,
        'esteri': esteri,
        'esteri_archive': esteri_archive,
        'docker_ravake': docker_ravake,
    }

    return names.get(name, dict())

# Default parameters for PPN, other dictionaries should override these parameters
# using dict.update() method.
defaults = {
    # Method selections
    "DOMAIN": "fmi", # See pystepsrc for valid data sources
    "OPTFLOW_METHOD": "lucaskanade",
    "FFT_METHOD": "pyfftw",
    "GENERATE_DETERMINISTIC": True,
    "GENERATE_ENSEMBLE": True,
    "GENERATE_UNPERTURBED": False,
    "REGENERATE_PERTURBED_MOTION": False,  # Re-calculate the perturbed motion fields used for pysteps nowcasting
    "VALUE_DOMAIN": "dbz",  # dbz or rrate
    # Z-R conversion parameters
    "ZR_A": 223.,
    "ZR_B": 1.53,
    # Nowcasting parameters
    "NUM_PREV_OBSERVATIONS": 3,
    "NOWCAST_TIMESTEP": 5,
    "MAX_LEADTIME": 120,
    "NUM_TIMESTEPS": None,
    "ENSEMBLE_SIZE": 25,
    "NUM_CASCADES": 6,
    "NUM_WORKERS": 6,
    "R_MIN": 0.1,
    "R_THRESHOLD": 0.1,
    "DBZ_MIN": -10,
    "DBZ_THRESHOLD": -10,
    "KMPERPIXEL": 1.0,
    "SEED": None,  # Default value in pysteps is None
    # Motion perturbation parameters
    # Set to VEL_PERT_KWARGS to `None` to use pysteps's default values
    "VEL_PERT_METHOD": "bps",
    "VEL_PERT_KWARGS": {
        # lucaskanade/fmi values given in pysteps.nowcasts.steps.forecast() method documentation
        "p_par": [2.20837526, 0.33887032, -2.48995355],
        "p_perp": [2.21722634, 0.32359621, -2.57402761],
    },
    # Storing parameters
    "FIELD_VALUES": "dbz",  # Store values as rrate or dbz
    "OUTPUT_PATH": None,  # None uses pystepsrc output path
    "OUTPUT_TIME_FORMAT": "%Y-%m-%d %H:%M:%S",
    "STORE_ENSEMBLE": True, # Write each ensemble member to output
    "STORE_UNPERTURBED": True,
    "STORE_DETERMINISTIC": True,  # Write det_fct to output
    "STORE_MOTION": True, # Write deterministic motion to output
    "STORE_PERTURBED_MOTION": True,  # Write motion for each ensemble member to output
    "SCALER": 100,
    "SCALE_ZERO": "auto",  # Value for "0" in scaled units. Set to "auto" or None for minimum value found before scaling
    # Logging parameters
    "WRITE_LOG": False,
    "LOG_LEVEL": logging.INFO,  # see logging module documentation for valid levels
    "LOG_FOLDER": "../logs",
}

# Add new parametrizations here
test = {
    "DOMAIN": "fmi_archive",
    "NOWCAST_TIMESTEP": 5,
    "MAX_LEADTIME": 30,
    "ENSEMBLE_SIZE": 5,
    "SEED": 0,

    "FIELD_VALUES": "rrate",
    "VALUE_DOMAIN": "rrate",

    "GENERATE_DETERMINISTIC": True,
    "GENERATE_ENSEMBLE": True,
    "GENERATE_UNPERTURBED": True,
    "REGENERATE_PERTURBED_MOTION": True,

    "STORE_DETERMINISTIC": True,
    "STORE_ENSEMBLE": True,
    "STORE_UNPERTURBED": True,
    "STORE_PERTURBED_MOTION": True,
    "STORE_MOTION": True,

    "LOG_LEVEL": logging.DEBUG,
    "WRITE_LOG": False,
}

esteri = {
    "DOMAIN": "fmi_realtime_ravake",
    "NOWCAST_TIMESTEP": 5,
    "MAX_LEADTIME": 90,
    "ENSEMBLE_SIZE": 15,
    "NUM_WORKERS": 30,
    "GENERATE_UNPERTURBED": True,
    "REGENERATE_PERTURBED_MOTION": True,  # Re-calculate the perturbed motion fields used for pysteps nowcasting
    "GENERATE_DETERMINISTIC": True,
    "OUTPUT_PATH": "/dev/shm/ppn",
    "STORE_MOTION": True,
    "STORE_ENSEMBLE": True,
    "STORE_DETERMINISTIC": True,
    "SEED": 20190823,
    "FIELD_VALUES": "dbz",
    "VALUE_DOMAIN": "dbz",
    "SCALER": 10,
    "WRITE_LOG": True,
    "LOG_LEVEL": logging.DEBUG,
    "LOG_FOLDER": "/var/tmp/log",
    "DBZ_MIN": -10,
    "DBZ_THRESHOLD": -10,
    "VEL_PERT_KWARGS": {
        # lucaskanade/fmi values given in pysteps.nowcasts.steps.forecast() method documentation
        "p_par": [0, 0, 0],
        "p_perp": [0, 0, 0],
    },
}

esteri_archive = {
    "DOMAIN": "fmi_archived_ravake",
    "NOWCAST_TIMESTEP": 5,
    "MAX_LEADTIME": 60,
    "ENSEMBLE_SIZE": 15,
    "NUM_WORKERS": 25,
    "GENERATE_DETERMINISTIC": True,
    "OUTPUT_PATH": "/dev/shm/ppn",
    "STORE_MOTION": True,
    "STORE_ENSEMBLE": True,
    "STORE_DETERMINISTIC": True,
    "SEED": 20190823,
    "FIELD_VALUES": "dbz",
    "VALUE_DOMAIN": "dbz",
    "SCALER": 10,
    "WRITE_LOG": True,
    "LOG_LEVEL": logging.DEBUG,
    "LOG_FOLDER": "/var/tmp/log",
    "DBZ_MIN": -6,
    "DBZ_THRESHOLD": -6,
}

docker_ravake = {
    "DOMAIN": "fmi_realtime_ravake_docker",
    "NOWCAST_TIMESTEP": 5,
    "MAX_LEADTIME": 90,
    "ENSEMBLE_SIZE": 15,
    "NUM_WORKERS": 30,
    "GENERATE_UNPERTURBED": True,
    "REGENERATE_PERTURBED_MOTION": True,  # Re-calculate the perturbed motion fields used for pysteps nowcasting
    "GENERATE_DETERMINISTIC": True,
    "OUTPUT_PATH": "~/fmippn-run/fmippn-run-and-distribution/output",
    "STORE_MOTION": True,
    "STORE_ENSEMBLE": True,
    "STORE_DETERMINISTIC": True,
    "SEED": 20190823,
    "FIELD_VALUES": "dbz",
    "VALUE_DOMAIN": "dbz",
    "SCALER": 10,
    "WRITE_LOG": True,
    "LOG_LEVEL": logging.DEBUG,
    "LOG_FOLDER": "~/fmippn-run/fmippn-run-and-distribution/logs",
    "DBZ_MIN": -10,
    "DBZ_THRESHOLD": -10,
    "VEL_PERT_KWARGS": {
        # lucaskanade/fmi values given in pysteps.nowcasts.steps.forecast() method documentation
        "p_par": [0, 0, 0],
        "p_perp": [0, 0, 0],
    },
}
